styx-scheduler-service/src/main/java/com/spotify/styx/docker/KubernetesDockerRunner.java

```diff
@@ -457,58 +458,70 @@ public void init() {
-    runningWorkflowInstances.forEach(workflowInstance -> stateManager.receiveIgnoreClosed(
-        Event.runError(workflowInstance, "No pod associated with this instance")));
+    // Emit errors for workflow instances that seem to be missing its pod
+    runningWorkflowInstances.forEach((workflowInstance, runState) -> {
+
+      // Is there a matching pod in the list? Bail.
+      if (workflowInstancesForPods.contains(workflowInstance)) {
+        return;
+      }
+
+      // The pod list might be stale so explicitly look for a pod using the execution ID.
+      final String executionId = runState.data().executionId().get();
+      final Pod pod = client.pods().withName(executionId).get();
+
+      // We found a pod? Bail.
+      if (pod != null) {
+        return;
+      }
+
+      // No pod found. Emit an error guarded by the state counter we are basing the error conclusion on.
+      stateManager.receiveIgnoreClosed(
+          Event.runError(workflowInstance, "No pod associated with this instance"), runState.counter());
+    });
   }
 
-  @VisibleForTesting
-  void pollPods() {
+  private void pollPods() {
     try {
       tryPollPods();
     } catch (Throwable t) {
       LOG.warn("Error while polling pods", t);
     }
   }
 
-  private synchronized void tryPollPods() {
-    final Set<WorkflowInstance> runningWorkflowInstances = getRunningWorkflowInstances();
+  @VisibleForTesting
+  synchronized void tryPollPods() {
+    // Fetch pods _before_ fetching all active states
     final PodList list = client.pods().list();
-    examineRunningWFISandAssociatedPods(runningWorkflowInstances, list);
+    final Map<WorkflowInstance, RunState> activeStates = stateManager.getActiveStates();
+    examineRunningWFISandAssociatedPods(activeStates, list);
 
     for (Pod pod : list.getItems()) {
       logEvent(Watcher.Action.MODIFIED, pod, list.getMetadata().getResourceVersion(), true);
       final Optional<WorkflowInstance> workflowInstance = readPodWorkflowInstance(pod);
       if (!workflowInstance.isPresent()) {
         continue;
       }
-      // do an extra lookup to get latest state
-      // this is crucial because after fetching all states in datastore, some states might
-      // have transitioned already; there could even be cases that we observe pods whose
-      // states where created after fetching the states.
-      final Optional<RunState> runState = lookupPodRunState(pod, workflowInstance.get());
-      if (runState.isPresent()) {
-        emitPodEvents(Watcher.Action.MODIFIED, pod, runState.get());
+
+      final RunState runState = activeStates.get(workflowInstance.get());
+      if (runState != null && isPodRunState(pod, runState)) {
+        emitPodEvents(Watcher.Action.MODIFIED, pod, runState);
         cleanupWithRunState(workflowInstance.get(), pod.getMetadata().getName());
       } else {
+        // The pod is stale as we fetched the active states _after_ listing all pods.
```

Comment:
There is a small issue here and not so harmful. After fetching all pods, while batch fetching active states, watcher receives pod main container termination event and transitions state (active state gets removed from datastore), so we will not get that active state. We would believe the pod is still running but no state, so we will delete it immediately not respecting graceful deletion period. We could read the pod again here but it would be too slow and not worth it.

Reply:
Currently `cleanup()` does read the pod again though. I guess we might want to optimize away that though, in which case this could become a problem.

Reply:
That reading has been removed in the sidecar PR I remember.

Reply:
nvm, that commits got lost.